trace-utils/src/test_utils/datadog_test_agent.rs

```diff
@@ -405,4 +434,85 @@ impl DatadogTestAgent {
+    /// reset the network connection even after ready states pass. Instead of adding arbitrary
+    /// sleeps to these tests, we can just retry the request. This function should not be used for
+    /// requests that are actually being tested, like sending payloads to the test agent. It should
+    /// only be used for requests to setup the test. Examples of when you would use this
+    /// function are for starting sessions or getting snapshot results.
+    ///
+    /// # Arguments
+    ///
+    /// * `req` - A `Request<Body>` representing the HTTP request to be sent.
+    /// * `max_retries` - An `i32` specifying the maximum number of retry attempts.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(Response<Incoming>)` - If the request succeeds. The status may or may not be
+    ///   successful.
+    /// * `Err(anyhow::Error)` - If all retry attempts fail or an error occurs during the request.
+    ///
+    /// ```
+    async fn agent_request_with_retry(
+        &self,
+        req: Request<Body>,
+        max_retries: i32,
+    ) -> anyhow::Result<Response<Incoming>> {
+        let mut attempts = 1;
+        let mut delay_ms = 100;
+        let (parts, body) = req.into_parts();
+        let body_bytes = body
+            .collect()
+            .await
+            .expect("Failed to collect body")
+            .to_bytes();
+        let mut last_response;
+
+        loop {
+            let client = hyper_migration::new_default_client();
+            let req = Request::from_parts(parts.clone(), Body::from_bytes(body_bytes.clone()));
+            let res = client.request(req).await;
+
+            match res {
+                Ok(response) => {
+                    if response.status().is_success() {
+                        return Ok(response);
+                    } else {
+                        println!(
+                            "Request failed with status code: {}. Request attempt {} of {}",
+                            response.status(),
+                            attempts,
+                            max_retries
+                        );
+                        last_response = Some(Ok(response));
+                    }
+                }
+                Err(e) => {
+                    println!(
+                        "Request failed with error: {}. Request attempt {} of {}",
+                        e, attempts, max_retries
+                    );
+                    last_response = Some(Err(e))
+                }
+            }
+
+            if attempts >= max_retries {
+                return match last_response {
+                    Some(Ok(resp)) => Ok(resp),
+                    Some(Err(e)) => Err(anyhow::Error::new(e)),
+                    None => Err(anyhow::anyhow!(
+                        "No response received after {} attempts",
+                        max_retries
+                    )),
+                };
```

Comment:
`last_repsonse` doesn't need to be an `Option`, because it's never assigned the None variant. And then you could refactor this to something like

```suggestion
                return Ok(last_response?) ;
```